nixos/modules/services/misc/guix/default.nix

```diff
@@ -0,0 +1,394 @@
+        user-specific profiles, cache, and state files.
+
+        ::: {.warning}
+        Changing it to something other than the default will rebuild the
+        package.
+        :::
+      '';
+      example = "/gnu/var";
+    };
+
+    publish = {
+      enable = mkEnableOption "substitute server for your Guix store directory";
+
+      generateKeyPair = mkOption {
+        type = types.bool;
+        description = ''
+          Whether to enable generating signing keys in {file}`/etc/guix` which
+          is required to initialize a substitute server. Otherwise, you can
+          just pass `--public-key=$FILE` and `--private-key=$FILE` in
+          {option}`services.guix.publish.extraArgs`.
+        '';
+        default = true;
+        example = false;
+      };
+
+      port = mkOption {
+        type = types.port;
+        default = 8181;
+        example = 8200;
+        description = ''
+          Port of the substitute server to listen to.
+        '';
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "nobody";
+        defaultText = "nobody";
+        description = ''
+          Name of the user to change once the server is up.
+        '';
+      };
+
+      extraArgs = mkOption {
+        type = with types; listOf str;
+        description = ''
+          Extra flags to pass to the substitute server.
+        '';
+        default = [];
+        example = [
+          "--compression=zstd:6"
+          "--discover=no"
+        ];
+      };
+    };
+
+    gc = {
+      enable = mkEnableOption "automatic garbage collection service for Guix";
+
+      extraArgs = mkOption {
+        type = with types; listOf str;
+        default = [ ];
+        description = ''
+          List of arguments to be passed to {command}`guix gc`.
+
+          When given no option, it will try to collect all garbage which is
+          often inconvenient so it is recommended to set some options.
+        '';
+        example = [
+          "--delete-generations=1m"
+          "--free-space=10G"
+          "--optimize"
+        ];
+      };
+
+      dates = lib.mkOption {
+        type = types.str;
+        default = "03:15";
+        example = "weekly";
+        description = ''
+          How often the garbage collection occurs. This takes the time format
+          from {manpage}`systemd.time(7)`.
+        '';
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable (lib.mkMerge [
+    {
+      environment.systemPackages = [ package ];
+
+      # Building the user pool for Guix.
+      users.users = guixBuildUsers cfg.nrBuildUsers;
+      users.groups.${cfg.group} = { };
+
+      # In case the user wants the subtitute server to advertise to the local
+      # network with the '--advertise' flag ('guix publish --advertise') and
+      # make use of the autodiscovery of the build daemon with '--discover=yes'
+      # ('guix-daemon --discover=yes').
+      services.avahi.enable = lib.mkDefault true;
```

Comment:
Why is this not dependend on some option?

Reply:
Guix uses Avahi (through guile-avahi) for the autodiscovery and network service advertising of the substitute server.

Updated the comment above it which is pretty bad (now that I re-read it).